akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndTailSpec.scala

```diff
@@ -90,6 +90,48 @@ class FlowPrefixAndTailSpec extends AkkaSpec {
       subscriber.expectSubscriptionAndComplete()
     }
 
+    "throw if tail is attempted to be materialized twice" in assertAllStagesStopped {
+      val futureSink = newHeadSink
+      val fut = Source(1 to 2).prefixAndTail(1).runWith(futureSink)
+      val (takes, tail) = Await.result(fut, 3.seconds)
+      takes should be(Seq(1))
+
+      val subscriber1 = TestSubscriber.probe[Int]()
+      tail.to(Sink(subscriber1)).run()
+
+      val subscriber2 = TestSubscriber.probe[Int]()
+      tail.to(Sink(subscriber2)).run()
+      subscriber2.expectSubscriptionAndError().getMessage should ===("Tail Source cannot be materialized more than once.")
+
+      subscriber1.requestNext(2).expectComplete()
+
+    }
+
+    "signal error if substream has been not subscribed in time" in assertAllStagesStopped {
+      val tightTimeoutMaterializer =
+        ActorMaterializer(ActorMaterializerSettings(system)
+          .withSubscriptionTimeoutSettings(
+            StreamSubscriptionTimeoutSettings(StreamSubscriptionTimeoutTerminationMode.cancel, 500.millisecond)))
+
+      val futureSink = newHeadSink
+      val fut = Source(1 to 2).prefixAndTail(1).runWith(futureSink)(tightTimeoutMaterializer)
+      val (takes, tail) = Await.result(fut, 3.seconds)
+      takes should be(Seq(1))
+
+      val subscriber = TestSubscriber.probe[Int]()
+      Thread.sleep(1000)
+
+      tail.to(Sink(subscriber)).run()(tightTimeoutMaterializer)
+      subscriber.expectSubscriptionAndError().getMessage should ===("Tail Source has not been subscribed in 500 milliseconds.")
+    }
+
+    "shut down main stage if substream is empty, even when not subscribed" in assertAllStagesStopped {
+      val futureSink = newHeadSink
+      val fut = Source.single(1).prefixAndTail(1).runWith(futureSink)
+      val (takes, tail) = Await.result(fut, 3.seconds)
+      takes should be(Seq(1))
+    }
```

Comment:
The above test shows why Http tests failed before. The body stream of Http entities were not consumed in several tests, since they were expected to be empty. The above test passes the `assertAllStagesStopped` criteria because if the substream is not materialized when master stream completion arrives, the main stream simply closes itself, leaving a tombstone in the child stage.